core/src/main/scala/InsufficientMatingMaterial.scala

```diff
@@ -28,15 +28,74 @@ object InsufficientMatingMaterial:
           }
       )
 
-  /*
+  /**
+   * Returns whether some square in `destinations` can be reached by a king moving from `startSquare`,
+   * while avoiding all squares in `forbidden`.
+   *
+   * `destinations` must not contain `startSquare`, or any square in `forbidden`.
+   */
+  def kingPathExists(startSquare: Square, destinations: Bitboard, forbidden: Bitboard): Boolean =
+    if destinations.intersects(forbidden) || destinations.contains(startSquare) then
+      throw IllegalArgumentException(
+        "`destinations` contains either the start square or some forbidden square"
+      )
+
+    var skip = forbidden
+    var frontier = startSquare.bb
+
+    while frontier.nonEmpty do
+      if frontier.intersects(destinations) then return true
+      skip |= frontier
+      frontier = frontier.fold(Bitboard.empty) { (newFrontier, sq) =>
+        newFrontier | (sq.kingAttacks & ~skip)
+      }
+    end while
```

Comment:
We almost never do procedural programming in scala. We almost never use `var`, `while` or `return`.
The only rare exception is for extreme performance optimization, after production profiling and benchmark testing.